### PR TITLE
Feature arg options

### PR DESCRIPTION
**Pull request by kmilo9999**
Sorry for this huge PR.
Please review this tool and let's have a conversation on how it works, how to deploy it with the web volume viewer and if new features need to be added.

### CODE REVIEW

File: progress_bar.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Print iterations progress
+def printProgressBar (iteration, total, prefix = '', suffix = '', decimals = 1, length = 100, fill = '█', printEnd = "\r"):
```
Comment (mirestrepo) [inline progress_bar.py:2]: stay consistent with function names - [PEP8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names) suggests snake case

Comment (mirestrepo) [reply progress_bar.py:2]: Also, once you switch to typer, this function can probably just go away - https://typer.tiangolo.com/tutorial/progressbar/

File: requirements.txt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+imagecodecs==2021.11.20
```
Comment (mirestrepo) [inline requirements.txt:1]: This file will be replaced by `pyproject.toml` once you are using poetry

File: tiff-2-png-converter.py [hunk 3]
```diff
@@ -8,189 +9,273 @@
 import tifffile  as tiff
 import numpngw
 from progress_bar import *
+from image_process_helper import *
```
Comment (mirestrepo) [inline tiff-2-png-converter.py:12]: Wildcard imports are generally discouraged. Consider importing just the functions you need or the module with as shorter alias to keep traceability

File: tiff-2-png-converter.py [hunk 4]
```diff
@@ -8,189 +9,273 @@
...
-             image_r[i] = img
-        elif "ch2" in name:
-             if verbose:
-                print("Load Image " + str(i) + " Channel 2 - "+ file_full_path )
-             image_g[i] = img
-        elif "ch3" in name:
-             if verbose:
-                print("Load Image " + str(i) + " Channel 3 - "+ file_full_path )
-             image_b[i] = img
+class Tiff_Processor():
```
Comment (mirestrepo) [inline tiff-2-png-converter.py:18]: Class names recommendation is ` CapWords` convention

File: tiff-2-png-converter.py [hunk 5]
```diff
@@ -8,189 +9,273 @@
...
+    equalize_histogram = False
+    equalize_histogram_method = None
+    gamma = False
+    gamma_val = 1
+    num_imgs = None
+    source = None
+    dest = None
+
+    @staticmethod
+    def add_arguments(parser):
```
Comment (mirestrepo) [inline tiff-2-png-converter.py:29]: This should become a little simpler when using typer

**Review by mirestrepo — CHANGES_REQUESTED** [review]
Besides the comments by @mcmcgrath13 to adopt poetry, typer and logging, I think for this scenario you should move away from the object oriented design and refactor the class `Tiff_Processor` into simple independent functions . You should end with one executable that clearly walks through the steps for converting.
Most functions need a docstring with a description, more specific naming, and I think they can be made less verbose. But I would suggest start one step at a time. Maybe:
1. Move to poetry
2. Move to typer + functional style
3. Move to logger + simplify directory walkthroughs  and checks, some of this simplification may happen as you change the explicit verbosity check for the logging builtin functionality

One final thought. Is this PR only changing the arguments? Because if the logic is un-touched, it may be simpler to make the propose bite-size PRs back to main instead of to this branch

File: image_process_helper.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+  imin, imax = img.min(), img.max()
+  img_c = img.copy()
+  img_c = ((img_c - imin) / (imax - imin)) ** igamma
+  img_c = img_c * (imax - imin) + imin
+
+  return img_c.astype(img_dtype)
+
+def equalize_image_histogram_opencv(img_array):
+    ##https://opencv24-python-tutorials.readthedocs.io/en/latest/py_tutorials/py_imgproc/py_histograms/py_histogram_equalization/py_histogram_equalization.html
```
[the remark this answers is not in the document]
Comment (broarr) [reply image_process_helper.py:16]: *docstring

**Comment by broarr** [comment]
@mirestrepo beat me to basically everything I was going to say. I'd also consider a code formatter and maybe a linter. I use `black` as a formatter and `flake8` as a linter and it's not terrible getting them working together. `flake8` might not be worth it for a script this size, but I think `black` is really helpful for projects of any size

File: progress_bar.py (hunk 1 above)
Comment (kmilo9999) [reply progress_bar.py]: I renamed it according to the naming convention.

File: image_process_helper.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kmilo9999) [reply image_process_helper.py]: I added explanation and link to docustring

File: tiff-2-png-converter.py (hunk 3 above)
Comment (kmilo9999) [reply tiff-2-png-converter.py:11]: removed star and added function used from that file

File: tiff-2-png-converter.py (hunk 4 above)
Comment (kmilo9999) [reply tiff-2-png-converter.py]: Renamed it following CapWords  convention

**Comment by kmilo9999** [comment]
I added blank to format code as suggested.

@mcmcgrath13 @mirestrepo   I would like to merge this working version to main, and then on another branch migrate it to typer.

**Comment by mcmcgrath13** [comment]
@kmilo9999 that sounds very reasonable to me.  Could you capture all of the comments/suggestions here that you want to do in follow on PRs and create issues for them?

File: image_process_helper.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
+import numpy as np
+import cv2
+
+
+def gamma_correction(img: np.ndarray, gamma: float = 1.0):
```
Comment (mcmcgrath13) [inline image_process_helper.py:5]: add docstring

File: tiff-2-png-converter.py (hunk 3 above)
Comment (mcmcgrath13) [reply tiff-2-png-converter.py:11]: I still the a star on line 11, could you update that one as well?

File: tiff-2-png-converter.py [hunk 8]
```diff
@@ -1,197 +1,358 @@
...
+        parser.add_argument("--num_imgs", help="Generate messages ouput ")
+
+        parser.add_argument(
+            "--equalize1",
+            help="equilize histogram in volume slices using opencv functionality",
+            action="store_true",
+        )
+
+        parser.add_argument(
+            "--equalize2",
```
Comment (mcmcgrath13) [inline tiff-2-png-converter.py:55]: these names should be more descriptive.  or perhaps make the type of equalizer a value passed to an `--equalize` flag

File: tiff-2-png-converter.py [hunk 9]
```diff
@@ -1,197 +1,358 @@
...
+            "-s",
+            "--source",
+            required=True,
+            help="Path to directory where the image sequence is located",
+        )
+        parser.add_argument(
+            "-d",
+            "--dest",
+            required=True,
+            help="Path to the directory and filename of the result merged image (png by default) will be saved. ",
```
Comment (mcmcgrath13) [inline tiff-2-png-converter.py:40]: could you put an example here? It's unclear to me if the filename is required or not

File: tiff-2-png-converter.py [hunk 10]
```diff
@@ -1,197 +1,358 @@
...
+            elif bits_per_sample != img.dtype:
+                print(
+                    "image "
+                    + file_full_path
+                    + "has a different sample type "
+                    + bits_per_sample
+                    + " expected: "
+                    + bits_per_sample
+                )
+                sys.exit("Check images bits per sample tag")
```
Comment (mcmcgrath13) [inline tiff-2-png-converter.py:119]: this should throw an error instead of printing and exiting

File: tiff-2-png-converter.py [hunk 11]
```diff
@@ -1,197 +1,358 @@
...
+
+            name, extension = os.path.splitext(file_full_path)
+            file_name = os.path.basename(name)
+            sequence_number = re.search("^p[0-9]{1,5}", file_name)
```
Comment (mcmcgrath13) [inline tiff-2-png-converter.py:123]: is this file naming method universal to all tiffs? (I know almost nothing about tiffs)

File: tiff-2-png-converter.py [hunk 12]
```diff
@@ -1,197 +1,358 @@
...
+
+        if len(image_r) != 0 or len(image_g) != 0 or len(image_b) != 0:
+            self.merge_RGB(
+                image_r, image_g, image_b, images, width, heigth, bits_per_sample
+            )
+
+        print("##Saving Image to: " + self.dest)
+        self.save_to_Image(images, width, heigth)
+
+    def interpolateData(self, img_array, width, heigth, num_images):
```
Comment (mcmcgrath13) [inline tiff-2-png-converter.py:178]:
```suggestion
    def interpolate_data(self, img_array, width, heigth, num_images):
```

File: tiff-2-png-converter.py [hunk 13]
```diff
@@ -1,197 +1,358 @@
...
+            else:
+                img_3d_array = np.dstack([img_3d_array, np.zeros((width, heigth))])
+        min_data = np.amin(np.amin(img_3d_array, axis=2))
+        max_data = np.amax(np.amax(img_3d_array, axis=2))
+        print("Minimum value in the whole array:%d" % (min_data))
+        print("Maximum value in the whole array:%d" % (max_data))
+        min_max_scale = np.arange(min_data, max_data)
+
+        # reshaped_array = np_array.reshape(4, 2)
+        # print("")
```
Comment (mcmcgrath13) [inline tiff-2-png-converter.py:197]: delete commented out code

File: tiff-2-png-converter.py [hunk 14]
```diff
@@ -1,197 +1,358 @@
...
+
+    def save_to_Image(self, volume, width, height):
```
Comment (mcmcgrath13) [inline tiff-2-png-converter.py:199]:
```suggestion
    def save_to_image(self, volume, width, height):
```

File: tiff-2-png-converter.py [hunk 15]
```diff
@@ -1,197 +1,358 @@
...
+        file = open(name + "_metadata", "a")
+        file.write("Width:" + str(width) + "\n")
+        file.write("Heigth:" + str(height) + "\n")
+        file.write("Depth:" + str(self.num_imgs) + "\n")
+        file.write("bps:" + str(image_out.dtype) + "\n")
+        file.write("amax:" + str(amax) + "\n")
+        file.close()
+        print("Image written to file-system : " + name + extension)
+
+    def merge_RGB(
```
Comment (mcmcgrath13) [inline tiff-2-png-converter.py:261]:
```suggestion
    def merge_rgb(
```

File: tiff-2-png-converter.py [hunk 16]
```diff
@@ -1,197 +1,358 @@
...
+        self, image_r, image_g, image_b, rgb_images, width, heigh, bits_per_sample
```
Comment (mcmcgrath13) [inline tiff-2-png-converter.py:262]:
```suggestion
        self, image_r, image_g, image_b, rgb_images, width, height, bits_per_sample
```

File: image_process_helper.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (kmilo9999) [reply image_process_helper.py]: I added a brief explanation of those two types of methods do to a docustring.

File: tiff-2-png-converter.py [hunk 17]
```diff
@@ -1,197 +1,358 @@
...
+                if self.equalize_histogram:
+                    g_channel = self.equalize_histogram_method(g_channel)
+            else:
+                g_channel = zero_image
+            if image_b[z] is not None:
+                b_channel = image_b[z]
+                if self.equalize_histogram:
+                    b_channel = self.equalize_histogram_method(b_channel)
+            else:
+                b_channel = zero_image
```
Comment (mcmcgrath13) [inline tiff-2-png-converter.py:289]: this chunk is pretty repetitive, could it be pulled out into a function? I wonder if passing around the three r/g/b images as a list of [r, g, b] might simplify some code

File: tiff-2-png-converter.py (hunk 8 above)
Comment (kmilo9999) [reply tiff-2-png-converter.py:55]: 1. All the functions in the image_process_helper file are being used. In this case, Is it necessary to be specific ?
`from image_process_helper import gamma_correction,equalize_image_histogram_opencv,equalize_image_histogram_custom`
vs
`from image_process_helper import *`

2. Let me rework this in the typer migration. I have to change the command args either way. I'll put it in the issue.

File: tiff-2-png-converter.py (hunk 9 above)
Comment (kmilo9999) [reply tiff-2-png-converter.py:40]: It is specify it that the arg is required.
let me rework this using typer.

File: tiff-2-png-converter.py (hunk 8 above)
Comment (mcmcgrath13) [reply tiff-2-png-converter.py:55]: Part of the reason this pattern is discouraged is for provenance tracking.  So maybe do `import image_process_helper as iph` then prepend your function calls with `iph.`?

Comment (kmilo9999) [reply tiff-2-png-converter.py:55]: Good call! I'm adding it to the latest changes

File: image_process_helper.py (hunk 7 above)
Comment (kmilo9999) [reply image_process_helper.py:5]: Done!

File: tiff-2-png-converter.py (hunk 10 above)
Comment (kmilo9999) [reply tiff-2-png-converter.py]: I Added a try/except + raise statements.

File: tiff-2-png-converter.py (hunk 13 above)
Comment (kmilo9999) [reply tiff-2-png-converter.py]: Done!

File: tiff-2-png-converter.py (hunk 14 above)
Comment (kmilo9999) [reply tiff-2-png-converter.py]: Done.

File: tiff-2-png-converter.py (hunk 15 above)
Comment (kmilo9999) [reply tiff-2-png-converter.py]: Done

File: tiff-2-png-converter.py (hunk 16 above)
Comment (kmilo9999) [reply tiff-2-png-converter.py]: done

File: tiff-2-png-converter.py (hunk 17 above)
Comment (kmilo9999) [reply tiff-2-png-converter.py]: I reworked the function.

**Review by mcmcgrath13 — COMMENTED** [review]
there's still one outstanding function rename to snake case.  I think it's fine to merge this as is, but I'd be sure to open issues for the following at a minimum:

* typer cli
  * args clarification
* logging
* making more functional
* poetry
* autoformatting - black
* linting - flake8
* CI for formatting and linting
* tests?

**Comment by kmilo9999** [comment]
Ok, I'll merge. 
The current changes use black to format python code.

**Pull request merged**